vbft governance support

Honglei-Cong opened the pull request:
1. governance contract support
2. fee splitting for deposited tokens
3. vbft chain support for governance
4. vbft bug fixings

Commit:
add oracle native contract

Signed-off-by: zhangmenghang <zhangmenghang@onchain.com>

Commit:
change oracle native contract file struct

Signed-off-by: zhangmenghang <zhangmenghang@onchain.com>

Commit:
add governance contract
init some functions

Signed-off-by: zhangmenghang <zhangmenghang@onchain.com>

Commit:
add gov contract

Commit:
add gov contract

Commit:
update governance contract

Commit:
update governance contract

Commit:
update governance contract

Commit:
update governance contract

xiemylogos committed:
contract deploy in genesis-block,load config from db

xiemylogos committed:
update chain config when committing contract txn

Honglei-Cong committed:
fix publicKey usage in vbft

keypair.PublicKey changed to interface, rather than struct.
change all *PublicKey to PublicKey

Commit:
update governance contract

Commit:
update governance contract

xiemylogos committed:
modify generate dops

(cherry picked from commit fa77f3a)

Commit:
fix data type

Commit:
update and bug fix

xiemylogos committed:
add batch find from leveldb

xiemylogos committed:
add batch find from leveldb

xiemylogos committed:
add genesis test function

xiemylogos committed:
add genesis test function

Commit:
fix init config

xiemylogos committed:
modify updatechainconfig

Commit:
add transfer func

xiemylogos committed:
remove maxtchangeview config

Honglei-Cong committed:
fix net merge issue, remote /net dir

Commit:
add node num check

xiemylogos committed:
add maxchangeview config

xiemylogos committed:
vbft update peer pool after new consensus network settled

xiemylogos committed:
modify key,add contract address

Commit:
fix bugs and add feesplit contract

Commit:
change big.int to uint64

Commit:
finish feeSplit contract

Commit:
finish feeSplit contract

Honglei-Cong committed:
make syncer incorporate with ledger syncer, update server start

Honglei-Cong committed:
fix statemgmt start-syncing request creation

Commit:
update governance and feSplit contract

Commit:
update governance and feeSplit

xiemylogos committed:
get peersinfo from db

xiemylogos committed:
get peersinfo from db

Commit:
add index table and fix split curve

xiemylogos committed:
genesis block add chainconfig

Commit:
fix split curve and unique index

Honglei-Cong committed:
fix making proposal in tx-ticker

fix making proposal after last block persisted in tx-ticker
update proposal backoff timeout, backoff with proposer ranking
update proposal fetching response processing

Commit:
fix error log and add register oracle node func

xiemylogos committed:
modify merge conflict

xiemylogos committed:
modify config parameter name

Commit:
reconstruct native contract

xiemylogos committed:
modify unit test

Honglei-Cong committed:
fix vbft start and add blkNum to bftAction

also fixed one block-syncing issue

Commit:
fix init bugs

xiemylogos committed:
rename file name

xiemylogos committed:
modify import file name

Commit:
add black list

Commit:
remove test config

Honglei-Cong committed:
fix peer syncer boundary handling

Honglei-Cong committed:
fix participant calculation

remove limit on calcuation, put limit on proposer/endoser checking

Honglei-Cong committed:
provides msg hash when msgpool.addMsg

remove msg-serialization when adding msg to msgpool
caller provide msg-hash when adding msg to msgpool

Honglei-Cong committed:
startNewRound improvement

1. add commit-done/endorse-done checking
2. fix fast forwarding, start new round if no consensus msg available
3. remove commit-done flag check in commit-msg processing
4. enhanced endorse-done/commit-done
5. fix error handling in processMsgEvent, log errors rather than quit
msg handler.
6. changed some error log to info

Honglei-Cong committed:
limit blockSyncReq to syncer

syncer is much slower than sync-detector, too much blockSyncReq
makes blockSyncReq channel full.

Honglei-Cong committed:
fix txpool timer

replace ticker with timer/reset.
why not ticker: ticker sends event continuously, when timer handler is
busy with some other processing, there'll some ticker events pending.
This is not txpool-ticker expected.

Honglei-Cong committed:
fix gofmt

Commit:
fix errors; update deserilize and serilize func; add some unit test

xiemylogos committed:
vbft load config from leveldb  use deserialize

Honglei-Cong committed:
fix empty consensus failure

1. fix empty block construction, add sys txns
2. update proposal participant rank calculation
3. fix canFastForward, not limit on syncing state

Honglei-Cong committed:
fix rebase code changes, remove TxHash from notify, remove bookkeepers from vbft

xiemylogos committed:
modify load config invoke uniform interface

Honglei-Cong committed:
update vrf calculation, remove proposer sig temply

Commit:
fix unVote peer bug

xiemylogos committed:
check txnpool use validheight

Honglei-Cong committed:
fix gofmt

laizy commented on consensus/vbft/block_pool.go:
```diff
@@ -623,7 +616,18 @@ func (pool *BlockPool) setBlockSealed(block *Block, forEmpty bool) error {
+			}
+		}
+
+		block.Block.Transactions = systemTxs
```
when dealing with `block, transaction` etc, we implicitly assume they are immutable after construction. here you have modified block's transactions, so you need explicitly recalculate transaction root and set block.Hash to nil.

Commit:
remove oracle temporary

Commit:
remove all fmt.Print

Honglei-Cong committed:
update vbft double proposal construction

1. update vbft block definition
2. change double-sig to double-block
3. update sig handlings in other modules accordingly

laizy commented on consensus/vbft/msg_pool.go:
```diff
@@ -49,27 +48,19 @@ func newConsensusRound(num uint64) *ConsensusRound {
-	}
-	if _, present := self.msgHashs[h]; present {
+func (self *ConsensusRound) hasMsg(msg ConsensusMsg, msgHash common.Uint256) (bool, error) {
+	if _, present := self.msgHashs[msgHash]; present {
```
need adjust the return value to `bool`

Honglei-Cong commented:
Some important fix needed for the pull request.
close it now.
